### ISSUE TITLE
Surprise error when the third argument of publish is "undefined"

### ISSUE THREAD

**Issue by salboaie** [NodeRedis/node_redis#874]
With the newest versions the redis client (doesn't happen with 0.12.1) , a surprise  error is generated when the third argument is explicitly undefined. Check the code: 

var redis = require("redis");
var client = redis.createClient();
client.publish("channel", "test"); //is fine
client.publish("channel", "test", function(){}); //is fine again

client.publish("channel", "test", undefined); //not fine!!!
/*  
 Error: ERR wrong number of arguments for 'publish' command
 at ReplyParser._parseResult (/home/salboaie/work/deps/node_modules/redis/lib/parser/javascript.js:42:20)
 at ReplyParser.execute (/home/salboaie/work/deps/node_modules/redis/lib/parser/javascript.js:141:28)
 at Socket.<anonymous> (/home/salboaie/work/deps/node_modules/redis/index.js:91:27)
 at Socket.emit (events.js:95:17)
 at Socket.<anonymous> (_stream_readable.js:765:14)
 at Socket.emit (events.js:92:17)
 at emitReadable_ (_stream_readable.js:427:10)
 at emitReadable (_stream_readable.js:423:5)
 at readableAddChunk (_stream_readable.js:166:9)
 at Socket.Readable.push (_stream_readable.js:128:10)
 */

Could you fix it? It will potentially break other libraries,etc

**Comment by BridgeAR** [comment]
A *fix* for this is easy, the question though is, if this should be supported at all. I thought about this before changing the behavior in that direction.
Because if you pass a explicit undefined, I'd say something is likely not working in your code as you expect it too. And since you passed the undefined, it's handled as any other value. But if you have such a use case you can handle this in your code, so you know everything is working as it should.

What do the other @NodeRedis/contributors think about this?

**Comment by salboaie** [comment]
Ok, but it could break legacy code that may be someone can't even touch. It really surprised me.. one day was working another day is mysteriously crushing for no reason... Please don't be evil with your fans :)     It could steal hours from other people lives for no reason... It is justified to have cases where we are  interested to know the results and cases where they are in the same program. 

var optionalFunction = undefined;
if(something){
   optionalFunction = function(){};
}
client.publish("channel", "testtesttesttesttesttesttesttesttes", optionalFunction);

**Comment by BridgeAR** [comment]
Well version 2.0 had a couple of breaking changes. That's also why it's called 2.0, as we use [Semver](http://semver.org/). I'll add another note about that change to the Breaking change part of the 2.0 release.
It's not that I want to be evil with you or anyone. But the behavior as it is right now is the one that I personally would expect and wish how my program should behave. It prevents people from not realizing that their code might not work as intended.

For node_redis the best way to receive the arguments is actually as an array. That way is the fasted and does not have this issue. So you could write this as: 
```js
var options = ['channel', 'test'];
if (something) {
    options.push(function() {});
}
client.publish(options);
// or
if (something) {
    client.publish('channel', 'test', optionalFunc);
} else {
    client.publish('channel', 'test');
}
```
So there are a couple possibilities how to write this and don't you think the benefit from telling others that their code is not working properly is higher than the negative effect by the BC?

**Comment by salboaie** [comment]
One thing to have in mind is that the stack of the error you throw is only from your code. Not very helpful...  Somebody receiving that error should check all the publish usages in all the libraries that  potentially are using your module.

**Comment by BridgeAR** [comment]
@salboaie the stack traces could be improved while in debug_mode, that's true. But that is an individual feature in my opinion.

I'll revert to the old behavior, since it seems to be an issue for more people.

**Comment by BridgeAR** [comment]
Fixed by 2ca42417bfd1b0221e98289459bf6799ad6913f4. I'll release 2.1.0 later tonight.
